activation/atxdb.go

```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
+// addAtxToNodeId inserts activation atx id by node
+func (db *ActivationDb) addAtxToNodeId(nodeId types.NodeId, atx *types.ActivationTx) error {
 	key := getNodeIdKey(nodeId)
-	ids, err := db.atxs.Get(key)
-	var atxs []types.AtxId
-	if err != nil {
-		//layer doesnt exist, need to insert new layer
-		ids = []byte{}
-		atxs = make([]types.AtxId, 0, 0)
-	} else {
-		atxs, err = decodeAtxIds(ids)
-		if err != nil {
-			return errors.New("could not get all atxs from database ")
-		}
-	}
-
-	// append needs to be sorted
-	atxs = append(atxs, atx.Id())
-	l := len(atxs)
-	if l > 1 {
-		lastAtx, err := db.getAtxUnlocked(atxs[l-2])
-		if err != nil {
-			return errors.New("could not get atx from database ")
-		}
-		if atx.Sequence < lastAtx.Sequence {
-			sort.Slice(atxs, func(i, j int) bool {
-				atx1, err := db.getAtxUnlocked(atxs[i])
-				if err != nil {
-					panic("inconsistent db")
-				}
-				atx2, err := db.getAtxUnlocked(atxs[j])
-				if err != nil {
-					panic("inconsistent db")
-				}
-
-				return atx1.Sequence < atx2.Sequence
-			})
-		}
-	}
-
-	w, err := encodeAtxIds(atxs)
+	err := db.atxs.Put(key, atx.Id().Bytes())
 	if err != nil {
-		return errors.New("could not encode layer atx ids")
+		return fmt.Errorf("failed to store ATX ID for node: %v", err)
 	}
-	return db.atxs.Put(key, w)
+	return nil
 }
 
-// GetNodeAtxIds returns the atx ids that were received for node nodeId
-func (db *ActivationDb) GetNodeAtxIds(nodeId types.NodeId) ([]types.AtxId, error) {
+// GetNodeLastAtxId returns the last atx id that was received for node nodeId
+func (db *ActivationDb) GetNodeLastAtxId(nodeId types.NodeId) (types.AtxId, error) {
 	key := getNodeIdKey(nodeId)
-	db.log.Debug("fetching atxIDs for node %v", nodeId.Key[:5])
+	db.log.Debug("fetching atxIDs for node %v", nodeId.ShortString())
 
-	db.RLock()
-	ids, err := db.atxs.Get(key)
-	db.RUnlock()
+	id, err := db.atxs.Get(key)
 	if err != nil {
-		return nil, err
-	}
-	atxs, err := decodeAtxIds(ids)
-	if err != nil {
-		return nil, err
+		return *types.EmptyAtxId, err
 	}
-	return atxs, nil
+	return types.AtxId{Hash: common.BytesToHash(id)}, nil
 }
 
-// GetEpochAtxIds returns all atx id's created in this epoch as seen by this node
-func (db *ActivationDb) GetEpochAtxIds(epochId types.EpochId) ([]types.AtxId, error) {
-	db.RLock()
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	db.RUnlock()
+// GetPosAtxId returns the best (highest layer id), currently known to this node, pos atx id
+func (db *ActivationDb) GetPosAtxId(epochId types.EpochId) (types.AtxId, error) {
+	idAndLayer, err := db.getTopAtx()
 	if err != nil {
-		return nil, err
+		return *types.EmptyAtxId, err
 	}
-	atxs, err := decodeAtxIds(ids)
-	if err != nil {
-		return nil, err
+	if idAndLayer.LayerId.GetEpoch(db.LayersPerEpoch) != epochId {
+		return types.AtxId{}, fmt.Errorf("current posAtx (epoch %v) does not belong to the requested epoch (%v)",
```

Comment:
I don't entirely agree with how you handle errors in this method. I would expect that if you failed to use the top atx as a positioning atx, you'll find and return another candidate from the atx db. However, you should probably wait to see if this is indeed an issue (at least for the sake of not hiding bugs) before we add this functionality. 
Unless you disagree with me, please add a todo and file an issue.

Reply:
I partly disagree. Let me explain:

The top ATX is the most advanced (in terms of layers) ATX we have. When we build our own ATX we'll do it in the current epoch, never retroactively, so the options are: 

1. The top ATX belongs to the current epoch = 👍🏻
2. The top ATX belongs to an older epoch = 👎🏻

If option 2 materialized there's no point to look for another ATX, since the newest one is too old.

What we can (read: should) do in that case is to subscribe to a newer, relevant ATX being processed. There's a TODO for that one level up, here (credit for that TODO goes to @antonlerner):

https://github.com/spacemeshos/go-spacemesh/blob/bd59b3a505b77d9d76aeab60fba8e5a8a2cd053e/activation/activation.go#L382-L391

Reply:
Just realized that the documentation on that method I referenced is out of date... Pushed an update now.